[SPARK-40360] ALREADY_EXISTS and NOT_FOUND exceptions

srielau opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?

This PR introduces the following error classes:

- PARTITIONS_ALREADY_EXIST
  Cannot ADD or RENAME TO partition(s) <partitionList> in table <tableName> because they already exist.
  Choose a different name, drop the existing partition, or add the IF NOT EXISTS clause to tolerate a pre-existing partition

- PARTITIONS_NOT_FOUND
  The partition(s) <partitionList> cannot be found in table <tableName>.
  Verify the partition specification and table name.
  To tolerate the error on drop use ALTER TABLE … DROP IF EXISTS PARTITION.

-  ROUTINE_ALREADY_EXISTS
  Cannot create the function <routineName> because it already exists.
  Choose a different name, drop or replace the existing function, or add the IF NOT EXISTS clause to tolerate a pre-existing function

- ROUTINE_NOT_FOUND
  The function <routineName> cannot be found. Verify the spelling and correctness of the schema and catalog.
  If you did not qualify the name with a schema and catalog, verify the current_schema() output, or qualify the name with the correct schema and catalog.
  To tolerate the error on drop use DROP FUNCTION IF EXISTS

- SCHEMA_ALREADY_EXISTS
  Cannot create schema <schemaName> because it already exists.
  Choose a different name, drop the existing schema, or add the IF NOT EXISTS clause to tolerate pre-existing schema

- SCHEMA_NOT_EMPTY
  Cannot drop a schema <schemaName> because it contains objects.
  Use DROP SCHEMA ... CASCADE to drop the schema and all its objects.

- SCHEMA_NOT_FOUND
  The schema <schemaName> cannot be found. Verify the spelling and correctness of the schema and catalog.
  If you did not qualify the name with a catalog, verify the current_schema() output, or qualify the name with the correct catalog.
  To tolerate the error on drop use DROP SCHEMA IF EXISTS.

- TABLE_OR_VIEW_ALREADY_EXISTS
  Cannot create table or view <relationName> because it already exists.
  Choose a different name, drop or replace the existing object, or add the IF NOT EXISTS clause to tolerate pre-existing objects

- TABLE_OR_VIEW_NOT_FOUND
  The table or view <relationName> cannot be found. Verify the spelling and correctness of the schema and catalog.
  If you did not qualify the name with a schema, verify the current_schema() output, or qualify the name with the correct schema and catalog.
  To tolerate the error on drop use DROP VIEW IF EXISTS or DROP TABLE IF EXISTS.
 
- TEMP_TABLE_OR_VIEW_ALREADY_EXISTS
  Cannot create the temporary view <relationName> because it already exists.
  Choose a different name, drop or replace the existing view,  or add the IF NOT EXISTS clause to tolerate pre-existing views.
  
Also (for JDBC data sources):

- INDEX_ALREADY_EXISTS
  Cannot create the index because it already exists. <message>.
 
- INDEX_NOT_FOUND
  Cannot find the index. <message>.

Some background:
* We use ROUTINE over FUNCTION to be future proof, if/when PROCEDUREs appear.
* We coarsify around TABLE_OR_VIEW_NOT_FOUND and TABLE_OR_VIEW_ALREADY_EXISTS (getting rid of dedicated reason as RENAME TABLE, etc.
* We combine PARTITION and PARTITIONS errors
* I use SCHEMA religiously. A debate can be had whether/ho/when to return NAMESPACE  

There is currently one failure caused by:

https://issues.apache.org/jira/browse/SPARK-40521
Hive based ALTER TABLE ADD PARTITION returns to many partitions in case of PARTITIONS_ALREADY_EXISTS.

### Why are the changes needed?
We want to convert all error to use the error-class framework


### Does this PR introduce _any_ user-facing change?

Yes, we are moving away from "free txt" and consolidate errors is error-classes.json.
This hardens the QA and code allowing us to improve error messages without breaking changes

### How was this patch tested?

Run existing QA suite

srielau committed:
[SPARK-40360] *_ALREADY_EXISTS and *_NOT_FOUND errors

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented on core/src/test/scala/org/apache/spark/SparkFunSuite.scala:
```diff
@@ -405,6 +413,18 @@ abstract class SparkFunSuite
       context: QueryContext): Unit =
     checkError(exception, errorClass, Some(errorSubClass), None, parameters, false, Array(context))
 
+  protected def checkErrorTableNotFound(exception: SparkThrowable,
+                                        tableName: String): Unit =
```
This function is very SQL specific, shall we put it in `QueryTest`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/identifiers.scala:
```diff
@@ -58,6 +58,16 @@ sealed trait CatalystIdentifier {
     }
   }
 
+  def parts: Seq[String] = {
```
`nameParts` is a more commonly used name in Spark.

srielau replied:
Done

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/package.scala:
```diff
@@ -130,6 +130,10 @@ package object util extends Logging {
     "`" + name.replace("`", "``") + "`"
   }
 
+  def quoteIdentifier(name: Seq[String]): String = {
```
`quoteNameParts`?

srielau replied:
Done

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -346,6 +346,18 @@
       }
     }
   },
+  "INDEX_ALREADY_EXISTS" : {
```
Where is it used?

MaxGekk replied:
It should appear somewhere in the source code:
```
$ find . -type f -print0|xargs -0 grep 'INDEX_ALREADY_EXISTS'
./core/target/scala-2.12/classes/error/error-classes.json:  "INDEX_ALREADY_EXISTS" : {
./core/src/main/resources/error/error-classes.json:  "INDEX_ALREADY_EXISTS" : {
./sql/core/src/main/scala/org/apache/spark/sql/jdbc/H2Dialect.scala:          // INDEX_ALREADY_EXISTS_1
```
but it is not in fact.

MaxGekk replied:
Maybe pass the error class here?
https://github.com/apache/spark/blob/288bdd28635d71623dfc6fe82dc1e3aab5fa6c63/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala#L84-L85

srielau replied:
Done, I wonder if this got lost in a merge...

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -346,6 +346,18 @@
+  },
+  "INDEX_NOT_FOUND" : {
+    "message" : [
+      "Cannot find the index. <message>."
```
What's the `message`? Can't you create sub-classes instead?

srielau replied:
@MaxGekk I would love to. However this one is a wrapper for JDBC. The message can be anything depending on the data source.
We will need a strategy on how to deal with these cases.

MaxGekk replied:
We convert `SQLException` in such way as I can see at:
https://github.com/apache/spark/blob/34d5272663ce4852ca5b2daa665983a321b42060/sql/core/src/main/scala/org/apache/spark/sql/jdbc/PostgresDialect.scala#L223-L232

Cann't we just pass the sqlState and `sqlException.getErrorCode` + vendor name. The info should be enough to identify the error, not?

MaxGekk replied:
> Can we do this later?

Yep, sure.

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -536,12 +563,71 @@
+    ],
+    "sqlState" : "42000"
+  },
+  "ROUTINE_NOT_FOUND" : {
```
Why do you name it `ROUTINE...` if SQL key word is `FUNCTION...`. The last one is even used in the error message. Could you rename to `FUNCTION...` for consistency, otherwise, please, explain the difference.

srielau replied:
Future proofing: Eventually we will add procedures.  The shared name for both is routine. This is also the name in the SQL Standard information schema.

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -536,12 +563,71 @@
+    ],
+    "sqlState" : "42000"
+  },
+  "SCHEMA_ALREADY_EXISTS" : {
```
Why not `NAMESPACE`?

srielau replied:
I list this as a concern in the description of the PR. IIUC @cloud-fan  PR to push errors into the catalog will enable us context specific errors.
The SQL Standard naming is: catalog.schema.(relation|type|..).
I was not able to discern when we go into the namespace codepath vs schema codepath.
CREATE SCHEMA gave a namespace error.

Ideally I'd say we add NAMEPACE_NOT_FOUND. NAMESPACE_ALREADY_EXISTS and raise the appropriate code. This should be a follow on PR.

srielau replied:
I have added these error classes. need to figure out how to drive them

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -536,12 +563,71 @@
   },
+  "TABLE_OR_VIEW_ALREADY_EXISTS" : {
+    "message" : [
+      "Cannot create table or view <relationName> because it already exists.",
```
`relationName` is new name, right? What's the old one, and why do you ignore it in exception.

srielau replied:
Not sure what you mean by old name. I don't think the local variables of the code are terribly relevant.
That being said relation is the generic name for table or views. The alternative is to call it tableOrViewName.

MaxGekk replied:
The error class comes from
https://github.com/apache/spark/blob/29e45528315c43336ba60ee1717ef5ec0c001c00/sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala#L809-L813
where `oldName` is not used anymore. So, you are going to give less info to users for the error troubleshooting.

MaxGekk commented on core/src/main/resources/error/error-classes.json:
```diff
@@ -536,12 +563,71 @@
+  },
+  "TEMP_TABLE_OR_VIEW_ALREADY_EXISTS" : {
+    "message" : [
+      "Cannot create the temporary view <relationName> because it already exists.",
```
The name is about temp **table** or view, but the error message says only about view.

srielau replied:
@MaxGekk  Because we do not have temp tables today, correct? The error class must be future proof. The text can evolve when we add the feature

MaxGekk replied:
Why not to create separate error classes? Or main error class like `ALREADY_EXISTS` + sub-classes per every relation type.

srielau replied:
We should be careful what we wish for here.. When a catalog runs into the primary key violation on the  "relation" scope it may not even know what the object is that it is trying to create. So we would force code to drag metadata around where it is not needed.
Either way in general we can can always suberror class state and objects, but s that really useful?
This classification may be mathematically cool, but I don't think it makes the error messages better.
In fact it can make them worse because we want application to CATCH these errors. If there are too many the exception handlers get lost (in fact I do have regrets abount approving the UNRESOLVED_COLUMN.WITH(OUT)RECOMMENDATION. IMHO we totally overshot and made it worse, just t avoid an empty list.

MaxGekk commented on core/src/test/scala/org/apache/spark/SparkFunSuite.scala:
```diff
@@ -316,14 +316,22 @@ abstract class SparkFunSuite
-      assert(actual.startIndex() === expected.startIndex(), "Invalid startIndex of a query context")
-      assert(actual.stopIndex() === expected.stopIndex(), "Invalid stopIndex of a query context")
-      assert(actual.fragment() === expected.fragment(), "Invalid fragment of a query context")
+    if (!queryContext.isEmpty) {
```
Please, remove this check. Test must reflect current imple of exception. If exceptions have already had query contexts, test should check them.

srielau replied:
Hmm, there was a reason for this.. I'll refresh my memory.

srielau replied:
OK, I have removed all that and checked the contexts (very painful!) and of dubious value IMHO.

MaxGekk commented on core/src/test/scala/org/apache/spark/SparkFunSuite.scala:
```diff
@@ -374,6 +382,19 @@ abstract class SparkFunSuite
     checkError(exception, errorClass, sqlState, parameters,
       matchPVals = true, Array(context))
 
+  protected def checkErrorTableNotFound(
```
Here is the core, right? which is not aware of any tables. Please, move it from here.

srielau replied:
@MaxGekk OK, where can I put it instead? This little function eliminates dozens of more complicated checkError invocations across many test files.
I added it because it's so pervasive

MaxGekk replied:
How about `QueryTest`, and pass a SQL statement or a plan into `checkErrorTableNotFound()`.

srielau replied:
There are two patterns:
`      val e = intercept[AnalysisException] {
        sql(s"ALTER TABLE $catalog.ns.no_tbl PARTITION (id=1) RENAME TO PARTITION (id=2)")
      }
      checkErrorTableNotFound(e, s"`$catalog`.`ns`.`no_tbl`",
        ExpectedContext(s"$catalog.ns.no_tbl", 12, 11 + s"$catalog.ns.no_tbl".length))
`
and
`
    val exc = intercept[NoSuchTableException] {
      catalog.loadTable(testIdent)
    }

    checkErrorTableNotFound(exc, testIdentQuoted)
`

Not sure how the API of that new shared function would look like.
@MaxGekk @cloud-fan

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala:
```diff
@@ -20,72 +20,105 @@ package org.apache.spark.sql.catalyst.analysis
-    this(s"Table or view '$table' already exists in database '$db'")
+    this(errorClass = "TABLE_OR_VIEW_ALREADY_EXISTS",
+      messageParameters = Map("relationName" ->
+        (quoteIdentifier(db) + "." + quoteIdentifier(table))))
```
Can't you used existing functions to form the name?

srielau replied:
I'm not aware of one. If you point me I can use it. At a minimum I'd have to build a sequence first. Just to unravel it.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -784,21 +784,17 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
 
   def renameTempViewToExistingViewError(oldName: String, newName: String): Throwable = {
-    new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1062",
```
If you remove this error classes, please, remove it from `error-classes.json` too.

srielau replied:
Done

srielau commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:
```diff
@@ -2319,20 +2307,19 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
-      messageParameters = Map(
-        "tablePath" -> tablePath,
-        "config" -> SQLConf.ALLOW_NON_EMPTY_LOCATION_IN_CTAS.key))
+      s"CREATE-TABLE-AS-SELECT cannot create table with location to a non-empty directory " +
+        s"${tablePath} . To allow overwriting the existing non-empty directory, " +
+        s"set '${SQLConf.ALLOW_NON_EMPTY_LOCATION_IN_CTAS.key}' to true.")
```
```suggestion
      errorClass = "_LEGACY_ERROR_TEMP_1241",
      messageParameters = Map(
        "tablePath" -> tablePath,
        "config" -> SQLConf.ALLOW_NON_EMPTY_LOCATION_IN_CTAS.key))

```

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalogSuite.scala:
```diff
@@ -1467,9 +1467,11 @@ abstract class SessionCatalogSuite extends AnalysisTest with Eventually {
+      checkError(e,
+        errorClass = "ROUTINE_ALREADY_EXISTS",
+        parameters = Map("routineName" -> "`temp1`"))
+            // Temporary function is overridden
```
Please, fix indentation.

srielau replied:
Done

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala:
```diff
@@ -45,14 +47,15 @@ trait AlterTableTests extends SharedSparkSession {
 
   test("AlterTable: table does not exist") {
     val t2 = s"${catalogAndNamespace}fake_table"
+    val quoted = UnresolvedAttribute.parseAttributeName(s"${catalogAndNamespace}table_name")
+      .map(part => quoteIdentifier(part)).mkString(".")
```
Does `toSQLId` do the same, no?

cloud-fan commented on connector/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/V2JDBCTest.scala:
```diff
@@ -154,21 +161,27 @@ private[v2] trait V2JDBCTest extends SharedSparkSession with DockerIntegrationFu
+    }
+    checkErrorTableNotFound(e,
+      UnresolvedAttribute.parseAttributeName(s"$catalogName.not_existing_table")
+        .map(part => quoteIdentifier(part)).mkString("."),
```
is it just `` `$catalogName`.`not_existing_table` ``?

srielau replied:
Catalog name itself can be catalog.schema which is why I need to do these contortions.

srielau commented:
@MaxGekk @cloud-fan 
Since the last review I have:
* eliminated tolerance for not having context and added all the contexts
* removed replaced template error codes
* addressed minor comments
* added namespace related error messages in error-classes.json

What I have NOT done and need help with:
* One remaining failure where I don't know how to generate the correct fragment in a multimode test. I have reached out to @gengliangwang to assist
* Pull checkErrorTableNotFound() out of SparkFunSuite. In fact, in a way, its now worse because of new overloading for context
* Generate NAMESPACE error codes

TBH this PR has been very expensive and impacted timely delivery of DBR docs.
So I would prefer to roll these two refactors into separate PRs driven by ICs.

srielau committed:
Fix SQLViewTestSuite

cloud-fan commented:
thanks, merging to master!

anchovYu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlreadyExistException.scala:
```diff
@@ -20,66 +20,112 @@ package org.apache.spark.sql.catalyst.analysis
 
 class IndexAlreadyExistsException(message: String, cause: Option[Throwable] = None)
-  extends AnalysisException(message, cause = cause)
+  extends AnalysisException(errorClass = "INDEX_NOT_FOUND",
```
Should this be INDEX_ALREADY_EXISTS?

MaxGekk replied:
This has been fixed in another PR as far as I remember. Please, check the master branch.
